Civi/Api4/Utils/CoreUtil.php

```diff
@@ -73,4 +73,34 @@ public static function getApiNameFromTableName($tableName) {
     return $entityName;
   }
 
+  /**
+   * Given an expression like "activity_date_time IS previous.month", rewrite to
+   * an equivalent expression "activity_date_time BETWEEN {start-of-previous-month} AND {end-of-previous-month}.
+   *
+   * @param string $fieldName
+   *   Ex: 'activity_date_time'
+   * @param string $operator
+   *   Ex: 'IS' or 'IS NOT'
+   * @param string $criteria
+   *   Ex: ['IS' => 'previous.month']
+   * @return array|NULL
+   *   Array(string $newOperator, mixed $newCriteria).
+   *   Ex: ['BETWEEN' => ['2020-05-01', '2020-06-01']]
+   */
+  public static function rewriteIsCriteria($fieldName, $operator, $criteria) {
+    $affirm = ($operator == 'IS');
+    if ($criteria === 'null' || $criteria === 'NULL') {
+      return $affirm ? ['IS NULL' => ''] : ['IS NOT NULL' => ''];
+    }
+
+    $relDateFilters = \CRM_Core_OptionGroup::values('relative_date_filters');
+    if (isset($relDateFilters[$criteria])) {
+      list ($dateFrom, $dateTo) = \CRM_Utils_Date::getFromTo($criteria, NULL, NULL);
+      $newOp = $affirm ? 'BETWEEN' : 'NOT BETWEEN';
+      return [$newOp => [\CRM_Utils_Date::mysqlToIso($dateFrom), \CRM_Utils_Date::mysqlToIso($dateTo)]];
+    }
```

Comment:
I don't think it's enough to compare the string input to the date filter option list and guess whether or not we're trying to use a relative date filter. What if an activity has the subject "this.year" and we want to search for that activity by subject?

I think this function needs at minimum the entire array of field metadata not just the field name, so it can first check the data type before assuming it's a date.

Reply:
> What if an activity has the subject "this.year" and we want to search for that activity by subject?

We came into this issue with a definition of the behavior for `['some_date', 'IS', 'this.year']`. However, we didn't explicitly discuss the behavior of `IS` in other scenarios. It could be:

1. Follow in the mold of SQL `IS` and perform basic equality matching on a user-string (same as `=`)
2. Unsupported - throw an error if one tries to use it differently.

There is definitely a learnability benefit of following the SQL `IS` semantics. However, as I think you're saying, things get ambiguous in this case... precisely because it's hard to distinguish a logical symbol (e.g. `"this.year"`) from a literal value (e.g. `"Hello world"`). SQL (as a DSL) can make affordances whenever they want to distinguish logical symbols and literal values (`null` vs `"null"`), but APIv4 operates within the tighter constraints of JSON/PHP-array ...  we can't introduce logical-symbols.

The approach in the current patch aims to resolve ambiguity by making a strict split: `=` matches literal-values, and `IS` matches on a logical symbol. So:

* `['subject', '=', 'this.year']` ==> Match the `subject` against the literal string `this.year`
* `['activity_date_time', 'IS', 'this.year']` ==> Match the `activity_date_time` using the logical symbol `this.year`
* Never the twain shall meet

My gut says that's overall simplest... i.e. no ambiguity, easy to implement.

But it breaks the intuition for people who know SQL... since SQL `IS` can take literal-value operands. We could either:

* Choose a different operator name (`=@` or `=$` or `$=` or somesuch?) to avoid confusion with SQL `IS`
* Generalize APIv4's approach to logical-symbols or pragmas.

    In this case, maybe we look at the other places where API currently uses logical-symbols or processing pragmas:

    * For chaining, the nested API call can use `$value.foo` to reference the prior result. (As it happens, this notation is ambiguous... because `'subject'=>'$value.foo'` could be a logical-symbol or literal-value)
    * For user/contact mapping, there's `@user:alice`. (As it happens, this notation is not problematic... because it's used in numerical contexts... and alphanumerical string necessarily requires evaluation/conversion.)

    Both of those suffer the problem of *obscurity*. The obscurity is amplified by the fact that you cannot write general docs for the API's expression-language -- each notation is bespoke to the field/value/use-case

    If we wanted, we could make the "relative date ranges" follow one of the previous examples closer. Ex: Bless `@` or`$` as the APIv4 sigil:

    | With `@` sigil in the value | With `$` sigil in the value |
    | -- | -- |
    | `['description' => '@value:display_name']` | `['description' => '$value.display_name']` |
    | `['id','=','@user:alice']` | `['id', '=', '$user.alice']` |
    | `['activity_date_time','=','@date:this.year']` | `['activity_date_time','=','$date.this.year']` |

    Any time you see an `@` or `$`, that would mean "the API preprocessor will do magic stuff on this thing." This is nice because it's consistent... and it can be adapted to many different uses within an API-call (get/create/update/delete; select/where/groupBy). But of course, then you have ambiguity for string-values that use `@` or `$`!

    We might resolve the ambiguity by moving the sigil to a different part of the array, eg

    | With `@` postfix in the key |
    | -- |
    | `['@description' => 'value.display_name]` |
    | `['@id', '=', 'user.alice']` |
    | `['@activity_date_time', '=', 'this.year']` |

    It feels a little funny, but I'm pretty sure resolves the ambiguity and works in many different use-cases.

So again... I think this PR (*using different operators for literal-values vs logical-symbols*) probably the simpler play. However, if the others seem higher value, then can we continue digging in...